appbuilder/services/npm-service.ts

```diff
@@ -0,0 +1,233 @@
+				this.$logger.trace(err.message);
+			}
+
+			return packageJsonContent;
+		}).future<any>()();
+	}
+
+	private hasTypesForDependency(packageName: string): IFuture<boolean> {
+		return (() => {
+			return !!this.getPackageJsonFromNpmRegistry(`${NpmService.TYPES_DIRECTORY}${packageName}`).wait();
+		}).future<boolean>()();
+	}
+
+	private buildNpmRegistryUrl(packageName: string, version: string): string {
+		return `${NpmService.NPM_REGISTRY_URL}/${packageName.replace("/", "%2F")}?version=${encodeURIComponent(version)}`;
+	}
+
+	private getPackageJsonContent(projectDir: string): IFuture<any> {
+		return (() => {
+			let pathToPackageJson = this.getPathToPackageJson(projectDir);
+
+			try {
+				return this.$fs.readJson(pathToPackageJson).wait();
+			} catch (err) {
+				if (err.code === "ENOENT") {
+					this.$errors.failWithoutHelp(`Unable to find ${this.$projectConstants.PACKAGE_JSON_NAME} in ${projectDir}.`);
+				}
+
+				throw err;
+			}
+
+		}).future<any>()();
+	}
+
+	private getPathToPackageJson(projectDir: string): string {
+		return path.join(projectDir, this.$projectConstants.PACKAGE_JSON_NAME);
+	}
+
+	private getPathToReferencesFile(projectDir: string): string {
+		return path.join(projectDir, this.$projectConstants.REFERENCES_FILE_NAME);
+	}
+
+	private installTypingsForDependency(projectDir: string, dependency: string): IFuture<void> {
+		return this.npmInstall(projectDir, `${NpmService.TYPES_DIRECTORY}${dependency}`, null, ["--save-dev", "--save-exact"]);
+	}
+
+	private generateReferencesFile(projectDir: string): IFuture<void> {
+		return (() => {
+			let packageJsonContent = this.getPackageJsonContent(projectDir).wait();
+
+			let pathToReferenceFile = this.getPathToReferencesFile(projectDir),
+				lines: string[] = [];
+
+			if (packageJsonContent && packageJsonContent.dependencies && packageJsonContent.dependencies[constants.TNS_CORE_MODULES]) {
+				let relativePathToTnsCoreModulesDts = `./${constants.NODE_MODULES_DIR_NAME}/${constants.TNS_CORE_MODULES}/${NpmService.TNS_CORE_MODULES_DEFINITION_FILE_NAME}`;
+
+				if (this.$fs.exists(path.join(projectDir, relativePathToTnsCoreModulesDts)).wait()) {
+					lines.push(this.getReferenceLine(relativePathToTnsCoreModulesDts));
+				}
+			}
+
+			if (packageJsonContent && packageJsonContent.devDependencies) {
+				_(packageJsonContent.devDependencies)
+					.keys()
+					.each(devDependency => {
+						if (this.isFromTypesRepo(devDependency)) {
+							let nodeModulesDirectory = path.join(projectDir, constants.NODE_MODULES_DIR_NAME);
+							let definitionFiles = this.$fs.enumerateFilesInDirectorySync(path.join(nodeModulesDirectory, devDependency),
+								(file, stat) => _.endsWith(file, constants.FileExtensions.TYPESCRIPT_DEFINITION_FILE) || stat.isDirectory(), { enumerateDirectories: false });
+
+							let defs = _.map(definitionFiles, def => {
+								return this.getReferenceLine(fromWindowsRelativePathToUnix(path.relative(nodeModulesDirectory, def)));
```

Comment:
Ohh, you ate right. I've changed this in the last minute and maybe I didn't test it very well. Feel free to fix it in case it really produces incorrect .abreferences.d.ts